File: Tester.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.concurrent.locks.ReentrantLock;

public class Tester {
    public static void main(String[] args) throws IOException, InterruptedException {
        System.out.println("Parent PID = " + ProcessHandle.current().pid());

        Process process = Runtime.getRuntime().exec("date");
        BufferedReader output = new BufferedReader(new InputStreamReader(process.getInputStream()));
        System.out.println(output.readLine());
        output.close();
        process.waitFor();
        System.out.println("First child PID = " + process.pid());

        process = Runtime.getRuntime().exec("date");
        output = new BufferedReader(new InputStreamReader(process.getInputStream()));
        System.out.println(output.readLine());
        output.close();
        process.waitFor();
        System.out.println("Second child PID = " + process.pid());

        ReentrantLock mutex = new ReentrantLock();
        Runnable runnable = new Runnable() {
            @Override
            public void run() {
                mutex.lock();
                System.out.println("Inside thread! Sleeping for 3 seconds...");
                try {
                    Thread.sleep(3000);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
                mutex.unlock();
            }
        };
        Thread t1 = new Thread(runnable);
        Thread t2 = new Thread(runnable);
        t1.start();
        t2.start();
        Thread.sleep(1000);
        System.out.println("Hello from main thread!");
        t1.join();
        t2.join();
    }
}

File: tester.c
#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <unistd.h>
#include <pthread.h>

pthread_mutex_t mutex;

void *thread_func(void *args)
{
    pthread_mutex_lock(&mutex);
    printf("Inside thread! Sleeping for 3 seconds...\n");
    sleep(3);
    pthread_mutex_unlock(&mutex);
    return NULL;
}

int main(void)
{
    printf("Parent PID = %d\n", getpid());

    switch (fork()) {
    case 0:
        execlp("date", "date", (char *) 0);
    default:
        pid_t child = wait(NULL);
        printf("First child PID = %d\n", child);
    }

    switch (fork()) {
    case 0:
        execlp("date", "date", (char *) 0);
    default:
        pid_t child = wait(NULL);
        printf("Second child PID = %d\n", child);
    }

    pthread_t t1, t2;
    pthread_mutex_init(&mutex, NULL);
    pthread_create(&t1, NULL, thread_func, NULL);
    pthread_create(&t2, NULL, thread_func, NULL);
    sleep(1);
    printf("Hello from main thread!\n");
    pthread_join(t1, NULL);
    pthread_join(t2, NULL);

    return 0;
}
